feat: Support generating presigned url for UploadPart operation

plu opened the pull request:
It adds methods to generate a presigned url for the `UploadPart` operation.

## Issue

Fixes #1808, #723

## Description of changes

It generates two new methods:

* `UploadPartInput.presignURL(config:expiration:)`
* `S3Client.presignedURLForUploadPart(input:expiration:)`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

plu commented:
I've tried to use a presigned URL to do a multipart upload, but without success so far.

```
curl -v -X PUT -T /tmp/bar.txt "https://bucketname.s3.eu-central-1.amazonaws.com/bar.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=REDACTED%2F20241030%2Feu-central-1%2Fs3%2Faws4_request&X-Amz-Date=20241030T103435Z&X-Amz-SignedHeaders=host&X-Amz-Expires=3600&X-Amz-Signature=REDACTED49a23b35439939b882a76daf9c5049b9e8d22"
```

Always gets me a 403 with:

```
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided.
...
```

plu commented:
❗This is still incomplete and the generated URL's are not working yet.

plu commented:
After [9e3244e](https://github.com/awslabs/aws-sdk-swift/pull/1809/commits/9e3244e57a2294935c1c6997262dee07466d73c2) the generated URL's are correct and uploading works :).

Will look into writing tests for this now.

plu commented on codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/UploadPartPresignedURLMiddleware.kt:
```diff
@@ -0,0 +1,45 @@
+                public typealias RequestType = SmithyHTTPAPI.HTTPRequest
+                
+                public func apply(input: InputType, builder: SmithyHTTPAPI.HTTPRequestBuilder, attributes: Smithy.Context) throws {
+                    if let partNumber = input.partNumber {
+                        let queryItem = Smithy.URIQueryItem(name: "partNumber".urlPercentEncoding(), value: Swift.String(partNumber).urlPercentEncoding())
+                        builder.withQueryItem(queryItem)
+                    }
+                    if let uploadId = input.uploadId {
+                        let queryItem = Smithy.URIQueryItem(name: "uploadId".urlPercentEncoding(), value: Swift.String(uploadId).urlPercentEncoding())
+                        builder.withQueryItem(queryItem)
+                    }
```
Wondering if there is a way to generate this.

sichanyoo replied:
Query item provider is already generated [here](https://github.com/awslabs/aws-sdk-swift/blob/ba7ed92d600ea3068fcc9a8b572dcb8e41d2df4d/Sources/Services/AWSS3/Sources/AWSS3/Models.swift#L16398-L16417). It handles adding input values with [@httpQuery trait ](https://smithy.io/2.0/spec/http-bindings.html#httpquery-trait) in Smithy model for the service to the request at runtime. 

The `UploadPartInput.queryItemProvider` gets used [here](https://github.com/awslabs/aws-sdk-swift/blob/ba7ed92d600ea3068fcc9a8b572dcb8e41d2df4d/Sources/Services/AWSS3/Sources/AWSS3/S3Client.swift#L7923) for normal request flow, and to add `QueryItemMiddleware` for the `presignURL` method, only change that needs to be made is probably adding the middleware codegen back for `UploadPart` in [here](https://github.com/awslabs/aws-sdk-swift/blob/ba7ed92d600ea3068fcc9a8b572dcb8e41d2df4d/codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/presignable/PresignableUrlIntegration.kt#L198-L207). I.e., this custom middleware is not necessary.

sichanyoo commented on codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/presignable/PresignableUrlIntegration.kt:
```diff
@@ -86,6 +89,9 @@ class PresignableUrlIntegration(private val presignedOperations: Map<String, Set
                 }
+                "com.amazonaws.s3#UploadPart" -> {
+                    renderMiddlewareClassForUploadPart(ctx, delegator, op)
+                }
```
As per previous comment, to use the already generated `queryItemProvider` for `UploadPartInput`, you won't need to add this custom middleware here anymore.

sichanyoo commented on codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/presignable/PresignableUrlIntegration.kt:
```diff
@@ -204,6 +210,10 @@ class PresignableUrlIntegration(private val presignedOperations: Map<String, Set
                 operationMiddlewareCopy.removeMiddleware(op, "OperationInputBodyMiddleware")
                 operationMiddlewareCopy.appendMiddleware(op, PutObjectPresignedURLMiddlewareRenderable())
             }
+            "com.amazonaws.s3#UploadPart" -> {
+                operationMiddlewareCopy.removeMiddleware(op, "OperationInputBodyMiddleware")
+                operationMiddlewareCopy.appendMiddleware(op, UploadPartPresignedURLMiddlewareRenderable())
+            }
```
This is where you could just add back the `QueryItemMiddleware` which has the name `OperationInputQueryItemMiddleware`; it's located [here](https://github.com/smithy-lang/smithy-swift/blob/10de568b01aa71134e22e6253525afb9d0f4f8cf/smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/middlewares/OperationInputQueryItemMiddleware.kt#L12).

sichanyoo commented on codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/presignable/PresignableUrlIntegration.kt:
```diff
@@ -267,9 +277,37 @@ class PresignableUrlIntegration(private val presignedOperations: Map<String, Set
         }
     }
 
+    private fun renderMiddlewareClassForUploadPart(codegenContext: SwiftCodegenContext, delegator: SwiftDelegator, op: OperationShape) {
+
+        val serviceShape = codegenContext.model.expectShape<ServiceShape>(codegenContext.settings.service)
+        val ctx = codegenContext.toProtocolGenerationContext(serviceShape, delegator)?.let { it } ?: run { return }
+
+        val opIndex = OperationIndex.of(ctx.model)
+        val inputShape = opIndex.getInput(op).get()
+        val outputShape = opIndex.getOutput(op).get()
+        val operationErrorName = MiddlewareShapeUtils.outputErrorSymbolName(op)
+        val inputSymbol = ctx.symbolProvider.toSymbol(inputShape)
+        val outputSymbol = ctx.symbolProvider.toSymbol(outputShape)
+        val outputErrorSymbol = Symbol.builder().name(operationErrorName).build()
+        val filename = ModelFileUtils.filename(ctx.settings, "${inputSymbol.name}+QueryItemMiddlewareForPresignUrl")
+        val headerMiddlewareSymbol = Symbol.builder()
+            .definitionFile(filename)
+            .name(inputSymbol.name)
+            .build()
+        delegator.useShapeWriter(headerMiddlewareSymbol) { writer ->
+            val queryItemMiddleware = UploadPartPresignedURLMiddleware(
+                inputSymbol,
+                outputSymbol,
+                outputErrorSymbol,
+                writer
+            )
+            MiddlewareGenerator(writer, queryItemMiddleware).generate()
+        }
+    }
+
```
Not needed if using preexisting codegen'd query item provider & runtime library's query item middleware

sichanyoo commented on codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/UploadPartPresignedURLMiddlewareRenderable.kt:
```diff
@@ -0,0 +1,31 @@
+package software.amazon.smithy.aws.swift.codegen.middleware
+
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.swift.codegen.SwiftWriter
+import software.amazon.smithy.swift.codegen.integration.ProtocolGenerator
+import software.amazon.smithy.swift.codegen.middleware.MiddlewareRenderable
+
+// This middleware renderer inserts a custom middleware named `UploadPartPresignedURLMiddleware`
+// into the operation stack.  It is only intended for use with S3 `UploadPart` and only when
+// generating a pre-signed URL.
+class UploadPartPresignedURLMiddlewareRenderable : MiddlewareRenderable {
+
+    override val name = "UploadPartPresignedURLMiddleware"
+
+    override fun render(
+        ctx: ProtocolGenerator.GenerationContext,
+        writer: SwiftWriter,
+        op: OperationShape,
+        operationStackName: String
+    ) {
+        super.renderSpecific(ctx, writer, op, operationStackName, "serialize")
+    }
+
+    override fun renderMiddlewareInit(
+        ctx: ProtocolGenerator.GenerationContext,
+        writer: SwiftWriter,
+        op: OperationShape
+    ) {
+        writer.write("\$L()", name)
+    }
+}
```
Not needed if using preexisting codegen'd query item provider & runtime library's query item middleware

sichanyoo reviewed (changes requested):
Thank you for submitting this PR. I added couple pointers & comments. 

Feel free to lmk if you'd rather have me take up the PR. I'm happy to do either (review your PR and have you drive it through to merge, or take up the work from you going forward).

plu commented:
> Feel free to lmk if you'd rather have me take up the PR. I'm happy to do either (review your PR and have you drive it through to merge, or take up the work from you going forward).

Thank you for your feedback, I'm planning to look into it today.

plu commented:
I hope I have addressed all your comments correctly. Once everything looks correct to you, I'd start looking into getting some test coverage on these changes. I have integrated and tested these changes in my Vapor project, and the generated presigned url's are still working correctly.

sichanyoo commented:
> I hope I have addressed all your comments correctly. Once everything looks correct to you, I'd start looking into getting some test coverage on these changes.

Everything looks good! I'll take a second look when integration test is added for the presigned URL for uploadPart 👍 

Just as a pointer, here's where you could add a test case for uploadPart: [S3PresigndRequestTests](https://github.com/awslabs/aws-sdk-swift/blob/main/IntegrationTests/Services/AWSS3IntegrationTests/S3PresignedRequestTests.swift). Note that because the `aws-sdk-swift/IntegrationTests/` has its own package manifest, its contents won't show in Xcode if you open `aws-sdk-swift`. You'll have to open `IntegrationTests` directory directly.

plu commented:
Decided to add a completely new file for the integration test, hope that makes sense. Also covered the other method that didn't have a test case yet (`presignedRequestForUploadPart`).

sichanyoo commented:
I triggered the CI run for the PR and it seems because this comes from a fork (either that or bc PR is created by someone outside of team, or perhaps both), it can't read the secrets tied to repo that's used to run integration tests in CI (see [here](https://github.com/awslabs/aws-sdk-swift/blob/9c9eb62b02496502150c5647d21540f48a7cad29/.github/workflows/integration-test.yml#L155)). I might have to put up a new PR with your changes to run CI & merge.

plu commented:
Thank you! You're right - a fork cannot access the secrets of the upstream repository for security reasons.

https://michaelheap.com/access-secrets-from-forks/

plu commented:
Thank you for your help and guidance! You can close this PR at any time in favor of the other one you're going to open.

sichanyoo commented:
The PR clone (https://github.com/awslabs/aws-sdk-swift/pull/1835) was merged. It should go out with a new release during today's release 👍
